src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -890,129 +970,117 @@ public enum ActivityIdFormat
-                    // We are OK with this because conceptually the class is still read-only.  
-                    ref string strRef = ref Unsafe.AsRef(in _asHexString);
-                    Interlocked.CompareExchange(ref strRef, SpanToHexString(new ReadOnlySpan<byte>(idPtr, sizeof(ulong) * 2)), null);
-                }
-            }
-            return _asHexString;
+            return _hexString ?? "00000000000000000000000000000000";
         }
 
         /// <summary>
         /// Returns the TraceId as a 32 character hexadecimal string.  
         /// </summary>
         public override string ToString() => ToHexString();
 
-        #region equality operators
-        public static bool operator ==(in ActivityTraceId traceId1, in ActivityTraceId traceId2)
+        public static bool operator ==(ActivityTraceId traceId1, ActivityTraceId traceId2)
         {
-            return traceId1._id1 == traceId2._id1 && traceId1._id2 == traceId2._id2;
+            return traceId1._hexString == traceId2._hexString;
         }
-        public static bool operator !=(in ActivityTraceId traceId1, in ActivityTraceId traceId2)
+        public static bool operator !=(ActivityTraceId traceId1, ActivityTraceId traceId2)
         {
-            return traceId1._id1 != traceId2._id1 || traceId1._id2 != traceId2._id2;
+            return traceId1._hexString != traceId2._hexString;
         }
         public bool Equals(ActivityTraceId traceId)
         {
-            return _id1 == traceId._id1 && _id2 == traceId._id2;
+            return _hexString == traceId._hexString;
         }
         public override bool Equals(object obj)
         {
             if (obj is ActivityTraceId traceId)
-                return _id1 == traceId._id1 && _id2 == traceId._id2;
+                return _hexString == traceId._hexString;
             return false;
         }
         public override int GetHashCode()
         {
-            return _id1.GetHashCode() + _id2.GetHashCode();
+            return ToHexString().GetHashCode();
         }
-        #endregion
 
-        #region private
         /// <summary>
         /// This is exposed as CreateFromUtf8String, but we are modifying fields, so the code needs to be in a constructor.  
         /// </summary>
         /// <param name="idData"></param>
         private ActivityTraceId(ReadOnlySpan<byte> idData)
         {
-            _id1 = 0;
-            _id2 = 0;
-            _asHexString = null;
             if (idData.Length != 32)
                 throw new ArgumentOutOfRangeException(nameof(idData));
-            Utf8Parser.TryParse(idData.Slice(0, 16), out _id1, out _, 'x');
-            Utf8Parser.TryParse(idData.Slice(16, 16), out _id2, out _, 'x');
+
+            Span<ulong> span = stackalloc ulong[2];
+
+            Utf8Parser.TryParse(idData.Slice(0, 16), out span[0], out _, 'x');
+            Utf8Parser.TryParse(idData.Slice(16, 16), out span[1], out _, 'x');
```

Comment:
What if these parsing operations fail?  They can't?

Reply:
it'll keep span[0] and span[1]  == 0, and overall traceId invalid according to the W3C spec.spec requires to generate a new random one in the case. I created a new issue https://github.com/dotnet/corefx/issues/38486, l'll fix it separately

Reply:
thanks for catching it @stephentoub !

Reply:
Addressed in commit https://github.com/dotnet/corefx/pull/37395/commits/95a307952a5b146077603b1e04761663328b433f